### PR TITLE
resource_host.go

### PR DESCRIPTION
**Pull request by pwong9595**
Add DiscoveryIP to host for host created in Day0 to setup iSCSI initiator

### CODE REVIEW

File: internal/resources/resource_host.go [hunk 1]
```diff
@@ -460,12 +485,17 @@ func resourceMetalHostRead(d *schema.ResourceData, meta interface{}) (err error)
...
 		}
 		volumeInfos = append(volumeInfos, vi)
+		discoveryIPs = append(discoveryIPs, i.DiscoveryIP)
 	}
 
 	if err := d.Set(hVolumeInfos, volumeInfos); err != nil {
 		return err
 	}
 
+	if err := d.Set(hDiscoveryIPs, removeDuplicateStr(discoveryIPs)); err != nil {
```
Comment (AnuBose) [inline internal/resources/resource_host.go:495]: How are you getting a  list of discoveryIps from Quake? A list of discovery Ips?
Should we  say which volume should use which discovery IP? I was thinking more of lines [key, value] for [Volume ID, disocvery IP]

File: internal/resources/resource_host.go [hunk 2]
```diff
@@ -415,6 +424,21 @@ func resourceMetalHostCreate(d *schema.ResourceData, meta interface{}) (err erro
 	return resourceMetalHostRead(d, meta)
 }
 
+func removeDuplicateStr(strSlice []string) []string {
```
Comment (dbozzato81) [inline internal/resources/resource_host.go:427]: This function can be moved to utils.go

**Comment by llabordehpe** [comment]
Why do we need this list of "discovery IPs" in the TF host resource ?
This information doesn't exist in the client API (https://github.com/HewlettPackard/hpegl-metal-client/blob/master/v1/pkg/client/model_host.go)
Are you planning on computing this list based on info from all the volumes attached to the host ?

**Review by dbozzato81 — COMMENTED** [review]
Please remember to update the `version` file as well as the `README.md` in the `examples` folder for the host resource with all new fields.

File: internal/resources/resource_host.go [hunk 3]
```diff
@@ -220,6 +221,14 @@ func hostSchema() map[string]*schema.Schema {
 			Optional:    true,
 			Description: "map of label name to label value for this host",
 		},
+		hDiscoveryIPs: {
+			Type:     schema.TypeList,
+			Optional: true,
+			Elem: &schema.Schema{
```
Comment (chaitra-mylarappachar) [inline internal/resources/resource_host.go:227]: Since this is not user configurable, 'Computed: true' is the right flag here.
https://www.terraform.io/plugin/sdkv2/schemas/schema-behaviors#computed

File: internal/resources/resource_host.go [hunk 4]
```diff
@@ -44,6 +44,7 @@ const (
 	hPortalCommOkay       = "portal_comm_okay"
 	hPwrState             = "power_state"
 	hLabels               = "labels"
+	hDiscoveryIPs         = "discovery_ips"
```
Comment (chaitra-mylarappachar) [inline internal/resources/resource_host.go:47]: since in the API, it is the inner field of 'iscsi', rename the label to reflect that iscsi_discovery_ip?

File: internal/resources/resource_host.go [hunk 5]
```diff
@@ -460,12 +485,17 @@ func resourceMetalHostRead(d *schema.ResourceData, meta interface{}) (err error)
 			vTargetIQN:   i.TargetIQN,
 		}
 		volumeInfos = append(volumeInfos, vi)
+		discoveryIPs = append(discoveryIPs, i.DiscoveryIP)
```
Comment (ron-saito) [inline internal/resources/resource_host.go:488]: could instead only append if unique.  would need to add to utils.go something like hpe-hcss/quake/pkg/util/slice_utils.go (AppendStrUnique).   I miss Java Sets...

File: internal/resources/resource_host.go (hunk 1 above)
Comment (pwong9595) [reply internal/resources/resource_host.go]: Was able to show VMaaS and Linux to use just 1 DiscoveryIP

File: internal/resources/resource_host.go (hunk 2 above)
Comment (pwong9595) [reply internal/resources/resource_host.go]: No need for the function anymore

File: internal/resources/resource_host.go (hunk 3 above)
Comment (pwong9595) [reply internal/resources/resource_host.go]: Change name to ISCSIDiscoveryIP matching quake and Computed: true

File: internal/resources/resource_host.go (hunk 5 above)
Comment (pwong9595) [reply internal/resources/resource_host.go]: No need for the function anymore

File: internal/resources/resource_host.go [hunk 6]
```diff
@@ -220,6 +221,12 @@ func hostSchema() map[string]*schema.Schema {
 			Optional:    true,
 			Description: "map of label name to label value for this host",
 		},
+		hISCSIDiscoveryIP: {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "List of iSCSI Discovery IP address.",
```
Comment (ron-saito) [inline internal/resources/resource_host.go:228]: is this a list?

File: internal/resources/resource_host.go [hunk 7]
```diff
@@ -44,6 +44,7 @@ const (
 	hPortalCommOkay       = "portal_comm_okay"
 	hPwrState             = "power_state"
 	hLabels               = "labels"
+	hISCSIDiscoveryIP     = "iscsi_discovery_ip"
```
Comment (llabordehpe) [inline internal/resources/resource_host.go:47]: Please use plural since it is a list:
   hISCSIDiscoveryIPs    = "iscsi_discovery_ips"

File: internal/resources/resource_host.go [hunk 8]
```diff
@@ -220,6 +221,12 @@ func hostSchema() map[string]*schema.Schema {
 			Optional:    true,
 			Description: "map of label name to label value for this host",
 		},
+		hISCSIDiscoveryIP: {
+			Type:        schema.TypeString,
```
Comment (llabordehpe) [inline internal/resources/resource_host.go:225]: It is a list, so use "Type:     schema.TypeList,", and you need to define the "Elem". Look at other TypeList above.

File: internal/resources/resource_host.go [hunk 9]
```diff
@@ -220,6 +221,12 @@ func hostSchema() map[string]*schema.Schema {
...
+			Optional:    true,
```
Comment (llabordehpe) [inline internal/resources/resource_host.go:226]: Remove "Optional" since it is "Computed".

File: internal/resources/resource_host.go (hunk 6 above)
Comment (pwong9595) [reply internal/resources/resource_host.go]: Fixed comment

File: internal/resources/resource_host.go (hunk 9 above)
Comment (pwong9595) [reply internal/resources/resource_host.go]: Removed optional

File: internal/resources/resource_host.go (hunk 8 above)
Comment (pwong9595) [reply internal/resources/resource_host.go:225]: No longer a list. I have shown VMaaS and CaaS team that one Discovery will work.

File: internal/resources/resource_host.go (hunk 7 above)
Comment (pwong9595) [reply internal/resources/resource_host.go]: No longer a list so back to just 1 address

File: internal/resources/resource_host.go [hunk 10]
```diff
@@ -481,6 +488,14 @@ func resourceMetalHostRead(d *schema.ResourceData, meta interface{}) (err error)
...
 		volumeInfos = append(volumeInfos, vi)
+
+		if discoveryIP == "" {
+			discoveryIP = i.DiscoveryIP
+		}
+	}
+
+	if err = d.Set(hISCSIDiscoveryAddress, discoveryIP); err != nil {
+		return fmt.Errorf("set discovery IP: %v", err)
 	}
```
Comment (ron-saito) [inline internal/resources/resource_host.go:499]: why is this needed when discovery IP is already in volumeInfos?

Comment (pwong9595) [reply internal/resources/resource_host.go:499]: Good point, either application has to loop through Volume Info or we have to...... Let me ask the users.

Comment (ron-saito) [reply internal/resources/resource_host.go:499]: if volume info slice is not empty, they would get the first entry to get the discoveryIP right?

Comment (pwong9595) [reply internal/resources/resource_host.go:499]: Actually, in this case which the host is there. In theory, they know which VolumeInfo to use to fetch the DiscoveryIP. I think pushing back to them is a better solution. Thanks
